File: draw_lines.lua
-- Draw Lines

hg = require("harfang")

hg.InputInit()
hg.WindowSystemInit()

res_x, res_y = 1280, 720
win = hg.RenderInit('Harfang - Draw Lines', res_x, res_y, hg.RF_VSync | hg.RF_MSAA4X)

line_count = 1000

shader = hg.LoadProgramFromFile('resources_compiled/shaders/white')

-- vertices
vtx_layout = hg.VertexLayout()
vtx_layout:Begin()
vtx_layout:Add(hg.A_Position, 3, hg.AT_Float)
vtx_layout:End()

vtx = hg.Vertices(vtx_layout, line_count * 2)

-- main loop
angle = 0

while not hg.ReadKeyboard():Key(hg.K_Escape) and hg.IsWindowOpen(win) do
	hg.SetViewClear(0, hg.CF_Color | hg.CF_Depth, hg.ColorI(64, 64, 64), 1, 0)
	hg.SetViewRect(0, 0, 0, res_x, res_y)

	vtx:Clear()
	for i = 0, line_count-1 do
		vtx:Begin(2 * i):SetPos(hg.Vec3(math.sin(angle + i * 0.005), math.cos(angle + i * 0.01), 0)):End()
		vtx:Begin(2 * i + 1):SetPos(hg.Vec3(math.sin(angle + i * -0.005), math.cos(angle + i * 0.005), 0)):End()
	end

	hg.DrawLines(0, vtx, shader)  -- submit all lines in a single call

	angle = angle + hg.time_to_sec_f(hg.TickClock())

	hg.Frame()
	hg.UpdateWindow(win)
end

hg.RenderShutdown()
hg.DestroyWindow(win)

File: draw_lines.go
package main

import (
	"math"
	"runtime"

	hg "github.com/harfang3d/harfang-go/v3"
)

func main() {
	hg.InputInit()
	hg.WindowSystemInit()

	resX, resY := int32(1280), int32(720)

	window := hg.RenderInitWithWindowTitleWidthHeightResetFlags("Harfang - Draw Lines", resX, resY, hg.RFVSync|hg.RFMSAA4X)

	lineCount := int32(1000)

	shader := hg.LoadProgramFromFile("resources_compiled/shaders/white")

	// vertices
	vtxLayout := hg.NewVertexLayout()
	vtxLayout.Begin()
	vtxLayout.Add(hg.APosition, 3, hg.ATFloat)
	vtxLayout.End()

	vtx := hg.NewVertices(vtxLayout, lineCount*2)

	// main loop
	angle := 0.0

	for !hg.ReadKeyboard().Key(hg.KEscape) && hg.IsWindowOpen(window) {
		hg.SetViewClearWithColDepthStencil(0, hg.CFColor|hg.CFDepth, hg.ColorI(64, 64, 64), 1.0, 0)
		hg.SetViewRect(0, 0, 0, uint16(resX), uint16(resY))

		vtx.Clear()
		for i := 0.0; i < float64(lineCount); i++ {
			vtx.Begin(int32(2 * i)).SetPos(hg.NewVec3WithXYZ(float32(math.Sin(angle+i*0.005)), float32(math.Cos(angle+i*0.01)), 0.0)).End()
			vtx.Begin(int32(2*i + 1)).SetPos(hg.NewVec3WithXYZ(float32(math.Sin(angle+i*-0.005)), float32(math.Cos(angle+i*0.005)), 0.0)).End()
		}

		hg.DrawLines(0, vtx, shader) // submit all lines in a single call

		angle = angle + float64(hg.TimeToSecF(hg.TickClock()))

		hg.Frame()
		hg.UpdateWindow(window)
		runtime.GC()
	}

	hg.RenderShutdown()
	hg.DestroyWindow(window)

}
